sdk/identity/Azure.Identity/src/EnvironmentCredential.cs

```diff
@@ -206,5 +165,139 @@ private async ValueTask<ExtendedAccessToken> GetTokenImplAsync(TokenRequestConte
+
+                if (tenantId != null)
+                {
+                    builder.Append(" AZURE_TENANT_ID");
+                }
+
+                if (clientId != null)
+                {
+                    builder.Append(" AZURE_CLIENT_ID");
+                }
+
+                if (clientSecret != null)
+                {
+                    builder.Append(" AZURE_CLIENT_SECRET");
+                }
+
+                if (username != null)
+                {
+                    builder.Append(" AZURE_USERNAME");
+                }
+
+                if (password != null)
+                {
+                    builder.Append(" AZURE_PASSWORD");
+                }
+
+                _unavailableErrorMessage = builder.Append(" ]").ToString();
+            }
+        }
+
+        private static Dictionary<string, string> ParseCredentialsFile(string filePath)
+        {
+            return JsonSerializer.Deserialize<Dictionary<string, string>>(File.ReadAllText(filePath));
+        }
+
+        private static async Task<Dictionary<string, string>> ParseCredentialsFileAsync(string filePath, CancellationToken cancellationToken)
+        {
+            using Stream s = File.OpenRead(filePath);
+            return await JsonSerializer.DeserializeAsync<Dictionary<string, string>>(s, null, cancellationToken);
+        }
+
+        private TokenCredential BuildCredentialForCredentialsFile(Dictionary<string, string> authData)
+        {
+            authData.TryGetValue("clientId", out string clientId);
+            authData.TryGetValue("clientSecret", out string clientSecret);
+            authData.TryGetValue("tenantId", out string tenantId);
+            authData.TryGetValue("activeDirectoryEndpointUrl", out string activeDirectoryEndpointUrl);
+
+            if (clientId == null || clientSecret == null || tenantId == null || activeDirectoryEndpointUrl == null)
+            {
+                StringBuilder builder = new StringBuilder("Malformed Azure SDK Auth file. The file should contain 'clientId', 'clientSecret', 'tenentId' and 'activeDirectoryEndpointUrl' values. Currently set values [");
+
+                if (clientId != null)
+                {
+                    builder.Append(" clientId");
+                }
+
+                if (clientSecret != null)
+                {
+                    builder.Append(" clientSecret");
+                }
+
+                if (tenantId != null)
+                {
+                    builder.Append(" tenantId");
+                }
+
+                if (tenantId != null)
+                {
+                    builder.Append(" activeDirectoryEndpointUrl");
+                }
+
+                builder.Append(" ]");
+
+                throw new Exception(builder.ToString());
+            }
+
+            return new ClientSecretCredential(tenantId, clientId, clientSecret, _pipeline.WithAuthorityHost(new Uri(activeDirectoryEndpointUrl)));
```

Comment:
Note that in cases where the SDK Auth File has a different Authority than what is configured on the pipeline we prefer the one in the file.  That feels like it's the correct behavior, but I wanted to call out the choice here.